inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_gather_nd.cpp

```diff
@@ -0,0 +1,103 @@
+
+    const auto dataShapeRank = static_cast<size_t>(dataShape.get_shape()[0]);
+    const auto indicesShapeRank = static_cast<size_t>(indicesShape.get_shape()[0]);
+
+    int64_t batchDims = gatherND->get_batch_dims();
+    VPU_THROW_UNLESS(batchDims >= 0 && batchDims < std::min(dataShapeRank, indicesShapeRank),
+                     "dynamicToStaticShapeGatherND: node {} has unsupported batch_dims which is expected to be"
+                     " in [0; min({}, {})), but {} was provided", gatherND->get_friendly_name(), dataShapeRank, indicesShapeRank, batchDims);
+
+    VPU_THROW_UNLESS(indicesShapeRank - batchDims >= 2,
+                     "dynamicToStaticShapeGatherND: node {} should have at least two non-batch dimension, {} provided",
+                     gatherND->get_friendly_name(), indicesShapeRank - batchDims);
+
+    std::vector<int64_t> indicesShapePart(indicesShapeRank - batchDims - 1);
+    std::iota(indicesShapePart.begin(), indicesShapePart.end(), batchDims);
+
+    std::shared_ptr<ngraph::Node> outputShape = std::make_shared<ngraph::opset5::Gather>(
+        indicesShape,
+        ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{indicesShapePart.size()}, indicesShapePart),
+        ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {0}));
+
+    const auto lastIndicesDim = gatherND->get_input_partial_shape(1)[indicesShapeRank - 1].get_length();
+    if (lastIndicesDim + batchDims < dataShapeRank) {
+        std::vector<int64_t> dataShapePart(dataShapeRank - batchDims - lastIndicesDim);
+        std::iota(dataShapePart.begin(), dataShapePart.end(), lastIndicesDim + batchDims);
+        const auto dataShapePartNode = std::make_shared<ngraph::opset5::Gather>(
+                dataShape,
+                ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{dataShapePart.size()}, dataShapePart),
+                ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {0}));
+        outputShape = std::make_shared<ngraph::opset5::Concat>(ngraph::NodeVector{outputShape, dataShapePartNode}, 0);
+    }
+
+    if (batchDims > 0) {
+        std::shared_ptr<ngraph::Node> batchDimsPart = std::make_shared<ngraph::opset5::Gather>(
+            indicesShape,
+            ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {0}),
+            ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {0}));
+
+        for (size_t i = 1; i < batchDims; i++) {
+            const auto batchDimI = std::make_shared<ngraph::opset5::Gather>(
+                indicesShape,
+                ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {i}),
+                ngraph::opset5::Constant::create(shapeElementType, ngraph::Shape{1}, {0}));
+
+            batchDimsPart = std::make_shared<ngraph::opset5::Multiply>(batchDimsPart, batchDimI);
```

Comment:
Is it possible to use Gather + ReduceProd?

Reply:
Good suggestion! Done.